lang/src/org/partiql/lang/ast/passes/StatementRedactor.kt

```diff
@@ -82,238 +73,256 @@ fun redact(statement: String,
+ * Redact [PartiqlAst.Expr.Lit]s not assigned with [safeFieldNames] to "***(Redacted)". Function calls that have an
+ * entry in [userDefinedFunctionRedactionConfig] will have their arguments redacted based on the redaction lambda.
  */
-private class StatementRedactionRewriter(
-        private val statement: String,
-        node: ExprNode,
-        private val userDefinedFunctionRedactionConfig: Map<String, UserDefinedFunctionRedactionLambda>
-) : AstRewriterBase() {
-    private val root = node
+private class StatementRedactionVisitor(
+    private val statement: String,
+    private val safeFieldNames: Set<String>,
+    private val userDefinedFunctionRedactionConfig: Map<String, UserDefinedFunctionRedactionLambda>
+) : PartiqlAst.Visitor() {
     private val sourceLocationMetaForRedaction = arrayListOf<SourceLocationMeta>()
     private val maskPattern = "***(Redacted)"
 
-    fun rewriteExprNode(): ExprNode {
-        return super.rewriteExprNode(root)
-    }
-
-    override fun rewriteSelectWhere(node: ExprNode): ExprNode {
-        redactOnField(node)
-        return super.rewriteSelectWhere(node)
+    override fun visitExprSelect(node: PartiqlAst.Expr.Select) {
+        if (node.where != null) {
+            redactExpr(node.where)
+        }
     }
 
-    override fun rewriteDataManipulationWhere(node: ExprNode): ExprNode {
-        redactOnField(node)
-        return super.rewriteDataManipulationWhere(node)
+    override fun visitStatementDml(node: PartiqlAst.Statement.Dml) {
+        if (node.where != null) {
+            redactExpr(node.where)
+        }
     }
 
-    override fun rewriteAssignment(node: Assignment): Assignment {
-        if (!validateSafeFieldNames(node.lvalue)) {
-            redactOnField(node.rvalue)
+    override fun visitAssignment(node: PartiqlAst.Assignment) {
+        if (!skipRedaction(node.target, safeFieldNames)) {
+            redactExpr(node.value)
         }
-        return super.rewriteAssignment(node)
     }
 
-    override fun rewriteDataManipulationOperationInsertValueOp(node: InsertValueOp): DataManipulationOperation {
+    override fun visitDmlOpInsertValue(node: PartiqlAst.DmlOp.InsertValue) {
         when (node.value) {
-            is Struct -> redactOnStructInInsertValueOp(node.value)
-            else -> redactOnField(node.value)
+            is PartiqlAst.Expr.Struct -> redactStructInInsertValueOp(node.value)
+            else -> redactExpr(node.value)
         }
-        return super.rewriteDataManipulationOperationInsertValueOp(node)
     }
 
-    private fun redactOnField(node: ExprNode) {
-        when (node) {
-            is NAry -> redactOnNAry(node)
-            is Literal -> redactOnLiteral(node)
-            is Seq -> redactOnSeq(node)
-            is Struct -> redactOnStruct(node)
-            is Typed -> redactOnTyped(node)
-            else -> { /* intentionally blank */ }
+    /**
+     * Returns the redacted [statement].
+     */
+    fun getRedactedStatement(): String {
```

Comment:
nit: I'd make this the top of the class so readers unfamiliar with the code can more easily see all of types of the properties defined in the constructor.

Reply:
Good idea. Moved to the top of the class